[FSDP] Ensure that customized non tensor optimizer state can be saved

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99214

The current logic does not actually handle all different non-tensor optimizer states correctly. This PR fixes the issue and adds a test.

This PR will solve https://github.com/pytorch/pytorch/issues/99079

Differential Revision: [D45021331](https://our.internmc.facebook.com/intern/diff/D45021331/)

fegin committed:
[FSDP] Ensure that customized non tensor optimizer state can be saved

The current logic does not actually handle all different non-tensor optimizer states correctly. This PR fixes the issue and adds a test.

Differential Revision: [D45021331](https://our.internmc.facebook.com/intern/diff/D45021331/)

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -1807,6 +1807,33 @@ def step():
+            osd = FSDP.optim_state_dict(model, optim, optim_state_dict=original_osd)
+            osd_to_load = FSDP.optim_state_dict_to_load(model, optim, osd )
+            for param_id, state in osd_to_load["state"].items():
+                # Addd customized value
```
nit: typo 😄 
```suggestion
                # Add customized value
```

awgu commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -1807,6 +1807,33 @@ def step():
+            step()
+            original_osd = deepcopy(optim.state_dict())
+            for param_id, state in original_osd["state"].items():
+                # Addd customized value
```
```suggestion
                # Add customized value
```

awgu commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -1807,6 +1807,33 @@ def step():
+                state["value2"] = None
+
+            osd = FSDP.optim_state_dict(model, optim, optim_state_dict=original_osd)
+            osd_to_load = FSDP.optim_state_dict_to_load(model, optim, osd )
```
nit: I wonder if lint will complain about this one
```suggestion
            osd_to_load = FSDP.optim_state_dict_to_load(model, optim, osd)
```

awgu reviewed (approved):
Thanks for the super fast fix! This looks good to me. I just left a few nits.

fegin committed:
Update on "[FSDP] Ensure that customized non tensor optimizer state can be saved"


The current logic does not actually handle all different non-tensor optimizer states correctly. This PR fixes the issue and adds a test.

Differential Revision: [D45021331](https://our.internmc.facebook.com/intern/diff/D45021331/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Ensure that customized non tensor optimizer state can be saved"


The current logic does not actually handle all different non-tensor optimizer states correctly. This PR fixes the issue and adds a test.

This PR will solve https://github.com/pytorch/pytorch/issues/99079

Differential Revision: [D45021331](https://our.internmc.facebook.com/intern/diff/D45021331/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

awaelchli reviewed (approved):
Thanks 🎉 !

fegin commented:
@pytorchbot merge
